### DISCUSSION TITLE
Change when validation functions are called

### DISCUSSION THREAD

**Discussion by Plochem** [eclipse-langium/langium#1503] [Q&A]
Is there a way to change when my custom validation functions are called? Right now, the appropriate validation function runs every time I modify the DSL, but I only want the function to be called when the user saves the file. Basically, is there something similar to CheckType from xtext?

**Comment by msujew** [comment]
Hey @Plochem,

we have a very similar concept as the `CheckType` annotation from Xtext (note that Xtext doesn't actually respect the annotation in the language server use case - at least last I checked in 2020). Whenever you register validations, you can select a category. We currently don't have an example available but you can take a look at our [test cases](https://github.com/eclipse-langium/langium/blob/75aa73e001848a5aaeb92e1648e1f07d6e7064f6/packages/langium/test/workspace/validation-registry.test.ts#L69-L82) to see how the API is intended to be used.

You can see an example how to call those validations in [here](https://github.com/eclipse-langium/langium/blob/75aa73e001848a5aaeb92e1648e1f07d6e7064f6/packages/langium/test/workspace/document-builder.test.ts#L178-L211). Note that you should wrap the `build` call in a `WorkspaceLock#write` call as it's not cancellable otherwise.

Since the validation categories don't have any attached semantics (except for `built-in` and `fast` running on every document change) you'll need to take care that the `build` calls is invoked on the `onDidSaveTextDocument` callback of the language server connection. See [this recent PR](https://github.com/eclipse-langium/langium/pull/1474) (not released yet, that will make it much easier to react on this event) for some more info on how to do that.
